payment/controllers.go

```diff
@@ -704,3 +705,102 @@ func VerifyCredential(service *Service) handlers.AppHandler {
+func HandleStripeWebhook(service *Service) handlers.AppHandler {
+	return handlers.AppHandler(func(w http.ResponseWriter, r *http.Request) *handlers.AppError {
+		ctx := r.Context()
+		// get logger
+		sublogger := logger(ctx).With().
+			Str("func", "HandleStripeWebhook").
+			Logger()
+
+		// get webhook secret from ctx
+		endpointSecret, err := appctx.GetStringFromContext(ctx, appctx.StripeWebhookSecretCTXKey)
+		if err != nil {
+			sublogger.Error().Err(err).Msg("failed to get stripe_webhook_secret from context")
+			return handlers.WrapError(
+				err, "error getting stripe_webhook_secret from context",
+				http.StatusInternalServerError)
+		}
+
+		b, err := requestutils.ReadWithLimit(r.Body, 65536)
+		if err != nil {
+			return handlers.WrapError(err, "error reading request body", http.StatusServiceUnavailable)
+		}
+
+		event, err := webhook.ConstructEvent(
+			b, r.Header.Get("Stripe-Signature"), endpointSecret)
+		if err != nil {
+			return handlers.WrapError(err, "error verifying webhook signature", http.StatusBadRequest)
+		}
+
+		// Handle invoice events
+		if event.Type == "invoice.updated" {
+			// Retrieve invoice from update events
+			var invoice stripe.Invoice
+			err := json.Unmarshal(event.Data.Raw, &invoice)
+			if err != nil {
+				return handlers.WrapError(err, "error parsing webhook JSON", http.StatusBadRequest)
+			}
+
+			subscription, err := service.scClient.Subscriptions.Get(invoice.Subscription.ID, nil)
+			if err != nil {
+				return handlers.WrapError(err, "error retrieving subscription", http.StatusInternalServerError)
+			}
+			orderID, err := uuid.FromString(subscription.Metadata["orderID"])
+			if err != nil {
+				return handlers.WrapError(err, "error retrieving orderID", http.StatusInternalServerError)
+			}
+
+			// If the invoice is paid set order status to paid, otherwise
+			if invoice.Paid {
+				err = service.Datastore.UpdateOrder(orderID, "paid")
+				if err != nil {
+					return handlers.WrapError(err, "error updating order status", http.StatusInternalServerError)
+				}
+				err = service.Datastore.UpdateOrderMetadata(orderID, "stripeSubscriptionId", subscription.ID)
+				if err != nil {
+					return handlers.WrapError(err, "error updating order metadata", http.StatusInternalServerError)
+				}
+				return handlers.RenderContent(r.Context(), "payment successful", w, http.StatusOK)
+			}
+			err = service.Datastore.UpdateOrder(orderID, "pending")
+			if err != nil {
+				return handlers.WrapError(err, "error updating order status", http.StatusInternalServerError)
+			}
+			err = service.Datastore.UpdateOrderMetadata(orderID, "stripeSubscriptionId", subscription.ID)
+			if err != nil {
+				return handlers.WrapError(err, "error updating order metadata", http.StatusInternalServerError)
+			}
+			return handlers.RenderContent(r.Context(), "payment failed", w, http.StatusOK)
+		}
+
+		// Handle subscription cancellations
+		if event.Type == "customer.subscription.deleted" {
```

Comment:
Would you make this into an enum?

Reply:
resolved: https://github.com/brave-intl/bat-go/pull/816/commits/36a7ef8bb20712a77ad9fafdda5767191643f439